Feature/hive source

rohitsinha54 opened the pull request:
Addressed comments from here: 

https://github.com/caskdata/hydrator-plugins/pull/43
- This does the output sink setting in OutputFormatProvider
- Handle nullable types
- Other minor comments.

Commit:
do output setting for HiveBatchSink in OutputFormatProvider

Commit:
Addressed minor comments

Commit:
Handle nullable types

nitinmotgi commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -49,12 +49,14 @@
    */
   public static HCatSchema toHiveSchema(Schema schema, HCatSchema tableSchema) {
     List<HCatFieldSchema> fields = Lists.newArrayList();
+    //TODO: Finalize a consistent way to whether do projecttion on source or not while reading and if needed
```
Let's say Hive tables has 5 fields A,B,C,D & E. If the hive source output schema contains only B,D & E then the record has to be projected.

nitinmotgi commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -88,7 +91,7 @@ private static PrimitiveTypeInfo getType(String name, Schema.Type category) {
       default:
         throw new IllegalArgumentException(String.format(
-          "Schema contains field '%s' with unsupported type %s", name, category.name()));
+          "Schema contains field '%s' with unsupported type %s", name, type));
```
What's the corrective action one can take if this problem happens and why would it happen.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -67,11 +69,12 @@ public static HCatSchema toHiveSchema(Schema schema, HCatSchema tableSchema) {
-    switch (category) {
+  private static PrimitiveTypeInfo getType(String name, Schema schema) {
+    Schema.Type type = schema.isNullable() ? schema.getNonNullable().getType() : schema.getType();
+    switch (type) {
```
are there plans to support record, array, and map types?

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -127,24 +130,25 @@ public static Schema toSchema(HCatSchema hiveSchema) {
       case SHORT:
       case INT:
-        return Schema.Type.INT;
+        return Schema.nullableOf(Schema.of(Schema.Type.INT)).getType();
```
do byte, char, and short fields work when we convert a HCatRecord into a StructuredRecord?

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -49,12 +49,14 @@
    */
   public static HCatSchema toHiveSchema(Schema schema, HCatSchema tableSchema) {
     List<HCatFieldSchema> fields = Lists.newArrayList();
+    //TODO: Finalize a consistent way to whether do projection on source or not while reading and if needed
```
how come this is a todo instead of doing it? Looks like it is doing a projection, but not checking types.

rohitsinha54 replied:
The inconsistent/unknown part is whether we should take schema(field and type) from the user or just the fields (like database plugin). After the above comment and @nitinmotgi  comment I am assuming we want to take the schema so I will add the type check part.

albertshau replied:
ok got it. On a somewhat related note, would it be possible to get the HCatSchema at configure time (configurePipeline) ? Or is it only possible at runtime (beforeSubmit)

rohitsinha54 replied:
With HCatalog to get  the schema of the table we need to first set the input by calling HCatInputFormat.setInput(Job, ....) This call sets properties in the hadoop configuration. Since, conf is not available during configurePipeline I don't think it will be possible.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -49,12 +49,14 @@
         // this field of the schema must exist in the table
         Preconditions.checkNotNull(tableSchema.get(name), "Missing field %s in table schema", name);
-        fields.add(new HCatFieldSchema(name, getType(name, field.getSchema().getType()), ""));
+        fields.add(new HCatFieldSchema(name, getType(name, field.getSchema()), ""));
```
Rather than creating the TypeInfo from the field schema, should just take whatever is already in the tableSchema since that is guaranteed to be correct.  Also need to perform validation to catch mismatches between the passed in Schema vs the actual HCatSchema

rohitsinha54 replied:
The inconsistent/unknown part is whether we should take schema(field and type) from the user or just the fields (like database plugin). After the above comments I am assuming we want to take the schema so I will add the type check part.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -127,8 +113,56 @@ public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, HCa
-    public SinkOutputFormatProvider() {
-      this.conf = new HashMap<>();
+    public SinkOutputFormatProvider(BatchSinkContext context, Job job, HiveSinkConfig config) throws IOException {
+      job.setOutputKeyClass(WritableComparable.class);
```
are these required? doesn't seem correct for multiple sinks.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -127,8 +113,56 @@ public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, HCa
+      return partitionValues;
+    }
+
+    private Map<String, String> getConfigurationAsMap (Configuration conf) {
```
remove empty space between 'AsMap (Configuration'

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -127,8 +113,56 @@ public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, HCa
+      MapDifference<String, String> mapDifference = Maps.difference(getConfigurationAsMap(originalConf),
+                                                                    getConfigurationAsMap(modifiedConf));
+      // new keys in the modified configurations
+      Map<String, String> stringStringMap = mapDifference.entriesOnlyOnRight();
```
nit: newEntries is a more descriptive name than stringStringMap

albertshau commented:
please include a test for a pipeline that writes to hive and something else.

Commit:
use field schema not main schema

Commit:
Do type check and addressed comments

albertshau commented on hive-plugins/src/main/java/co/cask/hydrator/plugin/batch/commons/HiveSchemaConverter.java:
```diff
@@ -52,9 +52,17 @@ public static HCatSchema toHiveSchema(Schema schema, HCatSchema tableSchema) {
+                     "The schema {} for field {} be used.", type, name, hiveType, hiveType, name);
+          type = hiveType;
+        }
+        fields.add(new HCatFieldSchema(name, type, ""));
```
could you add hCatFieldSchema directly or is this different?

albertshau commented on hive-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/HiveBatchSink.java:
```diff
@@ -127,8 +111,54 @@ public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, HCa
+      }
+      HCatOutputFormat.setSchema(modifiedConf, hiveSchema);
+      conf = getConfigurationDiff(originalConf, modifiedConf);
+      HiveSchemaStore.storeHiveSchema(context, config.dbName, config.tableName, hiveSchema);
```
this should happen in the beforeSubmit method rather than in this constructor. Usually don't want to do stuff that hits external systems in a constructor.

albertshau commented:
minor comments, otherwise lgtm assuming the char,byte,short stuff works, and assuming there is a jira open for unit tests

Commit:
Do proper conversion in the source while setting

Commit:
Addressed comments

Commit:
handle nulls properly during conversion

rohitsinha54 commented:
@albertshau : I  have addressed comment and improvised the conversion for source. Please have a look. I had to rebase with develop and force push to update the PR. Last 3 commits are new commits added after your last review.

We have a jira for unit test here : https://issues.cask.co/browse/CDAP-4224 I will add them in a separate PR. 
Thanks.

albertshau commented:
:+1:
